rosbag2_cpp/include/rosbag2_cpp/player_clock.hpp

```diff
@@ -0,0 +1,96 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_CPP__PLAYER_CLOCK_HPP_
+#define ROSBAG2_CPP__PLAYER_CLOCK_HPP_
+
+#include <chrono>
+#include <functional>
+#include <memory>
+
+#include "rcutils/time.h"
+#include "rosbag2_cpp/visibility_control.hpp"
+
+namespace rosbag2_cpp
+{
+class PlayerClockImpl;
+
+/**
+ * Used to control the timing of bag playback.
+ * This clock should be used to query times and sleep between message playing,
+ * so that the complexity involved around time control and time sources
+ * is encapsulated in this one place.
+ */
+class PlayerClock
+{
+public:
+  // Type representing the current time as according to the playback
+  typedef rcutils_time_point_value_t ROSTimePoint;
+  /**
+   * Type representing an arbitrary steady time, used to measure real-time durations
+   * This type is never exposed by the PlayerClock - it is only used as input for internal
+   * tracking.
+   */
+  typedef std::chrono::time_point<std::chrono::steady_clock> SteadyTimePoint;
+  typedef std::function<SteadyTimePoint()> NowFunction;
+
+  /**
+   * Constructor.
+   *
+   * \param starting_time: provides an initial offset for managing time
+   *    This will likely be the timestamp of the first message in the bag
+   * \param rate: Rate of playback, a unit-free ratio. 1.0 is real-time
+   * \param now_fn: Function used to get the current steady time
+   *   defaults to std::chrono::steady_clock::now
+   *   Used to control for unit testing, or for specialized needs
+   * \throws std::runtime_error if rate is <= 0
+   */
+  ROSBAG2_CPP_PUBLIC
+  PlayerClock(
```

Comment:
I was thinking to add a separate static factory constructor for sim time, or have the sim time version be a subclass with a different constructor. This constructor will eventually take a 'clock_publish_frequency' argument to handle the first two cases

Reply:
I've gone ahead and clarified how this will be implemented, by making `PlayerClock` a pure virtual interface and providing this logic as `TimeControllerClock`.